content/silver/ref/expr/pattern-matching.md

```diff
@@ -121,9 +143,35 @@ end
 ```
 The relative order of patterns for forwarding and non-forwarding productions is relevant for which pattern matches.
 In this example, the `d(_)` pattern will never be reached because any trees built by the `d` production will match the `a(_)` pattern.
+This is the same as placing a more general pattern before a more specific pattern in non-forwarding pattern matching.
 To match a forwarding production, place it at the beginning of the clauses.
 It is not an error to place it later, but it is unlikely the resulting semantics are the intended ones.
 
+Matching through forwarding occurs on a production basis, not a pattern basis.
+To see what we mean by this, suppose we have the following production:
+```
+abstract production foo
+top::Nt ::= c::Integer
+{
+  forwards to if c == 1 then foo(0) else bar(c);
+}
+```
+We try to match against the `foo` production with this `case` expression:
+```
+case x of
+| foo(0) -> -1
+| foo(a) -> a
+| _ -> 0
+end
+```
+If `x = foo(1)`, we match the second rule, `foo(a) -> a`.
+We try to match the `foo` pattern constructor of the first rule, which succeeds, then try to match the `0` pattern, which fails.
+We do not try forwarding `x` at this point to see if it will become `foo(0)`, instead moving on to the next match rule.
+This is what we mean by matching through forwarding occurring on a production basis:  we only forward to try to match a production pattern constructor if we could not match that constructor already, not if that constructor was matched but its children did not match the subpatterns.
```

Comment:
I think this sentence should come earlier - before using the phrase "production basis".  I did not know what that meant up on line 150 until I get down here after the example. I think it is better to state the general idea first and then use an example to further clarify.

Does "production basis" also mean that we can't get into non-terminating expansion of something like `foo`?  In attribute evalution this could happen if the condition `c == 1` was something else (maybe based on another child) and then it forwards to `foo(1)` instead of `foo(0)`.

Reply:
It prevents us from getting into a non-terminating expansion in forwarding if we are matching against that pattern.

Suppose we have this `foo` production:
```
abstract production foo
top::Nt ::= c::Integer
{
  forwards to if c == 0 then bar(c) else foo(c + 1);
}
```
We try to evaluate this `case` expression:
```
case foo(1) of
| foo(0) -> -1
| bar(a) -> a
| _ -> 0
end
```
The first pattern fails to match because `1` does not match `0` even though the top-level `foo` pattern matches, so we move to the next pattern without forwarding.  In the next pattern, we try to match `foo(1)` against `bar(a)`, which fails without matching the top-level `bar` pattern, so we forward and try again.  We now try to match `foo(2)` against `bar(a)`, and fail again for the same reason, so we forward and try again with `foo(3)`, then `foo(4)`, then `foo(5)`, and so on, never finding a match.

Because forwarding on a per-production basis only saves us from non-terminating expansion in very specific cases, I don't think it is worth mentioning.

Reply:
It is OK not to mention this if the description of the process makes that clear.